### PR TITLE
Support for Inter-grid meta-data

### PR DESCRIPTION
**Pull request by arporter**
This PR is for work done under #59 to support the necessary meta-data changes for inter-grid kernels.

### CODE REVIEW

**Comment by arporter** [comment]
All modified lines in dynamo0p3 are covered. However lines 100, 117, 128, 145 of parse.py are missed.

**Comment by arporter** [comment]
All modified lines in parse.py are now covered. I've deleted a check in `get_stencil()` that was not being exercised by the test suite since it was checking `len(metadata.args) > 0` but it was immediately after an `if ... and metadata.args:` which did exactly the same thing.

Once I've heard back from @christophermaynard (who needs to talk to Ben Shipway) about field vectors and implemented the appropriate thing then this will be ready to review.

**Comment by arporter** [comment]
Ben has advised that field-vectors should be kept as valid argument types as long as it is not too much work. I've therefore extended the test suite to check that we accept inter-grid kernels with field-vector arguments.

Ready for first review from @rupertford.

File: doc/dynamo0p3.rst [hunk 1]
```diff
@@ -601,12 +601,14 @@ accesses are found then PSyclone aborts.
 Stencil Metadata
 ^^^^^^^^^^^^^^^^
 
-Field metadata supports an optional 4th argument which specifies that
-the field is accessed as a stencil operation within the
-Kernel. Stencil metadata only makes sense if the associated field is
-read within a Kernel i.e. it only makes sense to specify stencil
-metadata if the first entry is ``GH_FIELD`` and the second entry is
-``GH_READ``.
+Field metadata supports an optional 4th argument which can be used
```
[the remark this answers is not in the document]
Comment (arporter) [reply doc/dynamo0p3.rst]: Good point. I've re-structured as you suggest.

File: doc/dynamo0p3.rst [hunk 2]
```diff
@@ -671,6 +673,41 @@ Below is an example of stencil information within the full kernel metadata.
 There is a full example of this distributed with PSyclone. It may
 be found in ``examples/dynamo0p3/eg5``.
 
+.. _dynamo0.3-intergrid-mdata:
+
+Inter-Grid Metadata
```
[the remark this answers is not in the document]
Comment (arporter) [reply doc/dynamo0p3.rst]: I've added inter-grid to the list of Kernel types in the introductory paragraph. I've also added a section on rules for that kernel type.

File: src/psyclone/dynamo0p3.py [hunk 3]
```diff
@@ -519,7 +532,8 @@ def __init__(self, arg_type):
                 "Internal error in DynArgDescriptor03.__init__, (2) should "
                 "not get to here")
         Descriptor.__init__(self, self._access_descriptor.name,
-                            self._function_space1, stencil=stencil)
+                            self._function_space1, stencil=stencil,
+                            mesh=mesh)
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/dynamo0p3.py]: I think it is. At the very least I suspect it is generic for all FE APIs (albeit we only have one FE api at the moment).

File: src/psyclone/dynamo0p3.py [hunk 4]
```diff
@@ -33,7 +31,7 @@
 # ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 # POSSIBILITY OF SUCH DAMAGE.
 # -----------------------------------------------------------------------------
-# Author R. Ford STFC Daresbury Lab
+# Authors R. Ford and A. Porter, STFC Daresbury Lab
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/dynamo0p3.py]: Oh yes. I've filled out the initials now.

File: src/psyclone/dynamo0p3.py [hunk 5]
```diff
@@ -743,6 +763,90 @@ def _validate(self, need_evaluator):
...
+                # so collect a list of other types
+                non_field_arg_types.add(arg.type)
+
+        mesh_list = mesh_dict.keys()
+        if not mesh_list:
+            # There are no meshes associated with any of the arguments so
+            # this is not an inter-grid kernel
+            return
+
+        if len(mesh_list) != 2:
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/dynamo0p3.py]: I've taken your pedantic point and raised it :-) Error test and message are now more robust.

File: src/psyclone/parse.py [hunk 6]
```diff
@@ -5,17 +5,54 @@
 # whose members are identified at https://puma.nerc.ac.uk/trac/GungHo/wiki
 # -----------------------------------------------------------------------------
 # Author L. Mitchell Imperial College
+
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/parse.py]: Done.

File: src/psyclone/parse.py [hunk 7]
```diff
@@ -36,15 +73,110 @@ def get_builtin_defs(api):
...
+    '''
+    Returns stencil_type and stencil_extent as a dictionary
+    object from stencil metadata if the metadata conforms to the
+    stencil(type[,extent]) format
+
+    :param metadata: Component of kernel meta-data stored as a node in the
+                     parser AST
+    :type metadata: :py:class:`psyclone.expression.FunctionVar`
+    :param list valid_types: List of valid stencil types (strings)
+    :raises ParseError: if the supplied meta-data is not a recognised
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/parse.py]: Oh yes - added.

**Comment by arporter** [comment]
I've addressed all of the reviewer's comments. Back to @rupertford for second review.

File: doc/dynamo0p3.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (rupertford) [reply doc/dynamo0p3.rst]: Much better, thanks.

File: src/psyclone/dynamo0p3.py [hunk 8]
```diff
@@ -459,17 +463,26 @@ def __init__(self, arg_type):
...
             self._function_space1 = arg_type.args[2].name
 
-            # The optional 4th argument is a stencil specification
+            # The optional 4th argument is either a stencil specification
+            # or a mesh identifier (for inter-grid kernels)
             if len(arg_type.args) == 4:
                 try:
-                    stencil = self._get_stencil(arg_type.args[3],
-                                                VALID_STENCIL_TYPES)
+                    from psyclone.parse import get_stencil, get_mesh
```
Comment (rupertford) [inline src/psyclone/dynamo0p3.py]: These changes are part of an __init__ routine which does not specify its arguments for phinx in a docstring. As this was implemented at about the same time as the rule change on documenting interfaces and as I'm actually not sure if we are meant to document __init__ routines, I will let this pass :-).

File: src/psyclone/dynamo0p3.py [hunk 9]
```diff
@@ -705,8 +719,14 @@ def __init__(self, ast, name=None):
...
-        ''' Check that the meta-data conforms to Dynamo 0.3 rules for a
-        user-provided kernel or a built-in '''
+        '''
+        Check that the meta-data conforms to Dynamo 0.3 rules for a
+        user-provided kernel or a built-in
+
+        :param bool need_evaluator: whether this kernel requires an
+                                    evaluator/quadrature
+        :raises: ParseError: if meta-data breaks the Dynamo 0.3 rules
+        '''
```
Comment (rupertford) [inline src/psyclone/dynamo0p3.py]: Good

File: src/psyclone/dynamo0p3.py [hunk 10]
```diff
@@ -743,6 +763,90 @@ def _validate(self, need_evaluator):
...
+        self._validate_inter_grid()
+
+    def _validate_inter_grid(self):
+        '''
+        Checks that the kernel meta-data obeys the rules for Dynamo 0.3
+        inter-grid kernels. If none of the kernel arguments has a mesh
+        associated with it then it is not an inter-grid kernel and this
+        routine silently returns.
+
+        :raises ParseError: if meta-data breaks inter-grid rules
```
Comment (rupertford) [inline src/psyclone/dynamo0p3.py]: This should be :raises:. I'll modify it myself to save time.

File: src/psyclone/tests/dynamo0p3_multigrid_test.py [hunk 11]
```diff
@@ -0,0 +1,200 @@
+# -----------------------------------------------------------------------------
+# BSD 3-Clause License
+#
+# Copyright (c) 2017, Science and Technology Facilities Council
+# (c) The copyright relating to this work is owned jointly by the Crown,
```
Comment (rupertford) [inline src/psyclone/tests/dynamo0p3_multigrid_test.py]: This file is owned by us. I'll change it myself to speed things up.

File: src/psyclone/tests/dynamo0p3_multigrid_test.py [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author R. Ford and A. R. Porter, STFC Daresbury Lab
```
Comment (rupertford) [inline src/psyclone/tests/dynamo0p3_multigrid_test.py]: Authors and R. W. Ford. I'll change these myself. Thanks for including me :-)

File: src/psyclone/tests/parse_test.py [hunk 13]
```diff
@@ -209,3 +209,28 @@ def test_duplicate_named_invoke():
...
+            "literal" in str(excinfo))
+    node = FunctionVar(["stencil()"])
+    with pytest.raises(ParseError) as excinfo:
+        _ = get_stencil(node, ["cross"])
+    assert ("Expecting format stencil(<type>[,<extent>]) but found stencil()"
+            in str(excinfo))
+    node = FunctionVar(["stencil", "cross"])
+    # Deliberately break the args member of node in order to trigger an
+    # internal error
+    node.args = [True]
```
Comment (rupertford) [inline src/psyclone/tests/parse_test.py]: Should this be done via monkeypatch? Not a big deal so I'll skip.

**Pull request merged**